Libro solution:

In [8]:
from random import random

In [9]:
# Función que se quiere minimizar
def funcion(x, y):
    sum1 = x**2 * (4 - 2.1*x**2 + x**4/3.0)
    sum2 = x*y
    sum3 = y**2 * (-4 + 4*y**2)
    return sum1 + sum2 + sum3

# Devuelve un número aleatorio dentro de un rango con
# distribución uniforme (proporcionada por random)
def aleatorio(inf, sup):
    return random() * (sup - inf) + inf

In [10]:
# Clase que representa una partícula individual y que facilita
# las operaciones necesarias
class Particula:

    # Algunos atributos de clase (comunes a todas las partículas)
    # Parámetros para actualizar la velocidad
    inercia = 1.4
    cognitiva = 2.0
    social = 2.0

    # Límites del espacio de soluciones
    infx = -10.0
    supx = 10.0
    infy = -10.0
    supy = 10.0

    # Factor de ajuste de la velocidad inicial
    ajusteV = 100.0

    # Crea una partícula dentro de los límites indicados
    def __init__(self):
        self.x = aleatorio(Particula.infx, Particula.supx)
        self.y = aleatorio(Particula.infy, Particula.supy)

        self.vx = aleatorio(Particula.infx / Particula.ajusteV,
                            Particula.supx / Particula.ajusteV)
        self.vy = aleatorio(Particula.infy / Particula.ajusteV,
                            Particula.supy / Particula.ajusteV)

        self.xLoc = self.x
        self.yLoc = self.y
        self.valorLoc = funcion(self.x, self.y)

    # Actualiza la velocidad de la partícula
    def actualizaVelocidad(self, xGlob, yGlob):
        cogX = Particula.cognitiva * random() * (self.xLoc - self.x)
        socX = Particula.social * random() * (xGlob - self.x)
        self.vx = Particula.inercia * self.vx + cogX + socX

        cogY = Particula.cognitiva * random() * (self.yLoc - self.y)
        socY = Particula.social * random() * (yGlob - self.y)
        self.vy = Particula.inercia * self.vy + cogY + socY

    # Actualiza la posición de la partícula
    def actualizaPosicion(self):
        self.x = self.x + self.vx
        self.y = self.y + self.vy

        # Debe mantenerse dentro del espacio de soluciones
        self.x = max(self.x, Particula.infx)
        self.x = min(self.x, Particula.supx)
        self.y = max(self.y, Particula.infy)
        self.y = min(self.y, Particula.supy)

        # Si es inferior a la mejor, la adopta como mejor
        valor = funcion(self.x, self.y)
        if valor < self.valorLoc:
            self.xLoc = self.x
            self.yLoc = self.y
            self.valorLoc = valor

In [11]:
# Mueve un enjambre de partículas durante las iteraciones indicadas.
# Devuelve las coordenadas y el valor del mínimo obtenido.
def enjambreParticulas(particulas, iteraciones, reduccionInercia):

    # Registra la mejor posición global y su valor
    mejorParticula = min(particulas, key=lambda p: p.valorLoc)
    xGlob = mejorParticula.xLoc
    yGlob = mejorParticula.yLoc
    valorGlob = mejorParticula.valorLoc

    # Bucle principal de simulación
    for _ in range(iteraciones):

        # Actualiza la velocidad y posición de cada partícula
        for p in particulas:
            p.actualizaVelocidad(xGlob, yGlob)
            p.actualizaPosicion()

        # Actualiza el mínimo global
        mejorParticula = min(particulas, key=lambda p: p.valorLoc)
        if mejorParticula.valorLoc < valorGlob:
            xGlob = mejorParticula.xLoc
            yGlob = mejorParticula.yLoc
            valorGlob = mejorParticula.valorLoc

        # Finalmente se reduce la inercia de las partículas
        Particula.inercia *= reduccionInercia

    return (xGlob, yGlob, valorGlob)

# Parámetros del problema
nParticulas = 10
iteraciones = 100
redInercia = 0.9

# Genera un conjunto inicial de partículas
particulas = [Particula() for _ in range(nParticulas)]

# Ejecuta el algoritmo del enjambre de partículas
print(enjambreParticulas(particulas, iteraciones, redInercia))

(0.08984200560735318, -0.7126563997462201, -1.0316284534898772)


Solución con programación estructurada

In [12]:
from random import random


In [13]:
# Función a minimizar
def funcion(x, y):
    return x**2 * (4 - 2.1*x**2 + x**4/3.0) + x*y + y**2 * (-4 + 4*y**2)

# Número aleatorio en rango
def aleatorio(inf, sup):
    return random() * (sup - inf) + inf

# Crear una partícula (como diccionario)
def crear_particula(infx, supx, infy, supy, ajusteV):
    x = aleatorio(infx, supx)
    y = aleatorio(infy, supy)

    vx = aleatorio(infx/ajusteV, supx/ajusteV)
    vy = aleatorio(infy/ajusteV, supy/ajusteV)

    return {
        "x": x,
        "y": y,
        "vx": vx,
        "vy": vy,
        "xLoc": x,
        "yLoc": y,
        "valorLoc": funcion(x, y)
    }

# Actualizar velocidad
def actualizar_velocidad(p, xGlob, yGlob, inercia, c1, c2):
    r1 = random()
    r2 = random()

    p["vx"] = (inercia * p["vx"] +
               c1 * r1 * (p["xLoc"] - p["x"]) +
               c2 * r2 * (xGlob - p["x"]))

    p["vy"] = (inercia * p["vy"] +
               c1 * r1 * (p["yLoc"] - p["y"]) +
               c2 * r2 * (yGlob - p["y"]))

# Actualizar posición
def actualizar_posicion(p, infx, supx, infy, supy):
    p["x"] += p["vx"]
    p["y"] += p["vy"]

    # Limitar espacio
    p["x"] = max(infx, min(p["x"], supx))
    p["y"] = max(infy, min(p["y"], supy))

    # Evaluar
    valor = funcion(p["x"], p["y"])
    if valor < p["valorLoc"]:
        p["xLoc"] = p["x"]
        p["yLoc"] = p["y"]
        p["valorLoc"] = valor

# PSO estructurado
def pso(n_particulas, iteraciones):
    # Parámetros
    infx, supx = -10.0, 10.0
    infy, supy = -10.0, 10.0
    inercia = 1.4
    c1 = 2.0
    c2 = 2.0
    ajusteV = 100.0
    reduccion = 0.9

    # Crear partículas
    particulas = [
        crear_particula(infx, supx, infy, supy, ajusteV)
        for _ in range(n_particulas)
    ]

    # Mejor global inicial
    mejor = min(particulas, key=lambda p: p["valorLoc"])
    xGlob, yGlob, valorGlob = mejor["xLoc"], mejor["yLoc"], mejor["valorLoc"]

    # Iteraciones
    for _ in range(iteraciones):
        for p in particulas:
            actualizar_velocidad(p, xGlob, yGlob, inercia, c1, c2)
            actualizar_posicion(p, infx, supx, infy, supy)

        mejor = min(particulas, key=lambda p: p["valorLoc"])
        if mejor["valorLoc"] < valorGlob:
            xGlob, yGlob, valorGlob = mejor["xLoc"], mejor["yLoc"], mejor["valorLoc"]

        inercia *= reduccion

    return xGlob, yGlob, valorGlob

In [15]:
# Ejecutar
resultado = pso(10, 100)
print(resultado)

(0.08984067336985153, -0.7126558740523861, -1.0316284534812972)


### Comparación de resultados

| Resultado | x | y | f(x, y) |
|----------|----------------------|----------------------|----------------------|
| 1 | 0.08984200560735318 | -0.7126563997462201 | -1.0316284534898772 |
| 2 | 0.08984165305545959 | -0.7126558379939526 | -1.0316284534869609 |

### Observación

Ambos resultados convergen al mismo mínimo global esperado.

### Conclusión

El algoritmo muestra consistencia en la solución, con variaciones muy pequeñas entre ejecuciones debido al componente aleatorio del PSO.